File: clients/cpp/model/Versioned.cpp
#include "Versioned.hpp"

Versioned::Versioned() { }
Versioned::Versioned(std::unordered_map<int, UnitAction> inner) : inner(inner) { }
Versioned Versioned::readFrom(InputStream& stream) {
    Versioned result;
    size_t innerSize = stream.readInt();
    result.inner = std::unordered_map<int, UnitAction>();
    result.inner.reserve(innerSize);
    for (size_t i = 0; i < innerSize; i++) {
        int innerKey;
        innerKey = stream.readInt();
        UnitAction innerValue;
        innerValue = UnitAction::readFrom(stream);
        result.inner.emplace(std::make_pair(innerKey, innerValue));
    }
    return result;
}
void Versioned::writeTo(OutputStream& stream) const {
    stream.write(43981);
    stream.write((int)(inner.size()));
    for (const auto& innerEntry : inner) {
        stream.write(innerEntry.first);
        innerEntry.second.writeTo(stream);
    }
}
std::string Versioned::toString() const {
    return std::string("Versioned") + "(" +
        "TODO" + 
        ")";
}

File: clients/go/model/versioned.go
package model

import "io"
import . "aicup2019/stream"

type Versioned struct {
    Inner map[int32]UnitAction
}
func NewVersioned(inner map[int32]UnitAction) Versioned {
    return Versioned {
        Inner: inner,
    }
}
func ReadVersioned(reader io.Reader) Versioned {
    result := Versioned {}
    InnerSize := ReadInt32(reader)
    result.Inner = make(map[int32]UnitAction)
    for i := int32(0); i < InnerSize; i++ {
        var InnerKey int32
        InnerKey = ReadInt32(reader)
        var InnerValue UnitAction
        InnerValue = ReadUnitAction(reader)
        result.Inner[InnerKey] = InnerValue
    }
    return result
}
func (value Versioned) Write(writer io.Writer) {
    WriteInt32(writer, 43981)
    WriteInt32(writer, int32(len(value.Inner)))
    for InnerKey, InnerValue := range value.Inner {
        WriteInt32(writer, InnerKey)
        InnerValue.Write(writer)
    }
}
